hints.hs

```diff
@@ -0,0 +1,309 @@
+    | up == '$' = True
+    | down == '$' = True
+    | left  == '$' = True
+    | right == '$' = True
+    | up == '*' = True
+    | down == '*' = True
+    | left  == '*' = True
+    | right == '*' = True
+    | otherwise    = False
+  where up    = getCharacter b (pr-1,pc)
+        down  = getCharacter b (pr+1,pc)
+        left  = getCharacter b (pr,pc-1)
+        right = getCharacter b (pr,pc+1)
+
+-- given push,
+-- returns updated state if box pushed to unoccupied or goal space
+tryPush :: Push -> Maybe State
+tryPush ((State b ps (pr,pc) lvl), (br,bc), (r,c))
+    | (dest1 == '$') && (dest2 == ' ') = Just (State (pushToUnoccupied b (pr,pc) (br,bc) (r,c)) (State b ps (pr,pc) lvl) (br,bc) lvl)
+    | (dest1 == '$') && (dest2 == '.') = Just (State (pushToGoal b (pr,pc) (br,bc) (r,c)) (State b ps (pr,pc) lvl) (br,bc) lvl)
+    | (dest1 == '*') && (dest2 == ' ') = Just (State (pushGoalToUnoccupied b (pr,pc) (br,bc) (r,c)) (State b ps (pr,pc) lvl) (br,bc) lvl)
+    | (dest1 == '*') && (dest2 == '.') = Just (State (pushGoalToGoal b (pr,pc) (br,bc) (r,c)) (State b ps (pr,pc) lvl) (br,bc) lvl)
+    | otherwise                        = Nothing
+  where dest1 = getCharacter b (br,bc)
+        dest2 = getCharacter b (r,c)
+
+-- given list of pushes,
+-- returns list of (push, path)
+toPushesTodo :: [Push] -> [State] -> [(Push, [State])]
+toPushesTodo lp ls = map (\p -> (p, ls)) lp
+
+-- given a list of states,
+-- returns a list of adjacent pushes
+toPushes :: [State] -> [Push]
+toPushes ls = foldr (\s p -> (toPush s) ++ p) [] ls
+
+-- given a state,
+-- returns a list of adjacent pushes (max 4 cardinal directions)
+toPush :: State -> [Push]
+toPush (State b ps (pr,pc) lvl) = filter isBoxAt [up, down, left, right]
+  where up    = ((State b ps (pr,pc) lvl), (pr-1,pc), (pr-2,pc))
+        down  = ((State b ps (pr,pc) lvl), (pr+1,pc), (pr+2,pc))
+        left  = ((State b ps (pr,pc) lvl), (pr,pc-1), (pr,pc-2))
+        right = ((State b ps (pr,pc) lvl), (pr,pc+1), (pr,pc+2))
+
+-- given push,
+-- returns True if there is a box to push, False otherwise
+isBoxAt :: Push -> Bool
+isBoxAt ((State b ps (pr,pc) lvl), (br,bc), (r,c))
+    | x == '$'  = True
+    | x == '*'  = True
+    | otherwise = False
+  where x = getCharacter b (br,bc)
+
+
+{- Deadlock functions -}
+
+-- given a board,
+-- returns a list of deadlock coordinates (the board becomes unsolvable if any box is moved to the coords)
+findDeadSquares :: Board -> [Coordinates]
+findDeadSquares b = map coords (filter (isDeadSquare b) (findPotentialDeadSquares b))
+
+
+-- given a board,
+-- returns a list of potential dead squares (all ' ' and '@')
+findPotentialDeadSquares :: Board -> [DeadSquare]
+findPotentialDeadSquares b = [(setDeadSquare b (r,c)) | r <- [0..(length b - 1)],
+                                                        c <- [0..(length (head b) - 1)],
+                                                        (getCharacter b (r,c) == ' ') || (getCharacter b (r,c) == '@')]
```

Comment:
This just makes a list of empty squares (and the player when not starting on a goal), since all other squares can't be dead squares.

- '#' : box can't be moved to a wall (walls remain walls)
- '$' : assuming that boxes' starting positions are okay (no unsolvable levels given)
- '.', '*', '+' : boxes can be moved to a goal (it may be a problem in some cases for a particular box to get moved to a goal and become stuck, when solutions only have a different box, but for the purposes of this check it's easier to just allow the search alg to continue)

I can add a comment if it helps.

Reply:
ah okay, makes sense